components/Autocomplete/index.jsx

```diff
@@ -0,0 +1,258 @@
+
+const isMobile = window.outerWidth < 550;
+
+export default class Autocomplete extends React.Component {
+  static propTypes = {
+    data: PropTypes.object,
+    disabled: PropTypes.bool,
+    error: PropTypes.string,
+    id: PropTypes.string.isRequired,
+    inputStyle: PropTypes.string,
+    label: PropTypes.string.isRequired,
+    labelHidden: PropTypes.bool,
+    name: PropTypes.string.isRequired,
+    noSuggestionText: PropTypes.string,
+    onGetSuggestions: PropTypes.func,
+    placeholder: PropTypes.string,
+    query: PropTypes.object,
+    required: PropTypes.bool,
+    requiredLabel: PropTypes.string,
+    scrollOffset: PropTypes.number,
+    scrollTo: PropTypes.bool,
+    submitOnEnter: PropTypes.bool,
+    value: PropTypes.string
+  };
+
+  static defaultProps = {
+    error: null,
+    inputStyle: 'inline',
+    noSuggestionText: 'No results found',
+    placeholder: '',
+    query: {},
+    scrollTo: false
+  }
+
+  state = {
+    showError: false,
+    showNoSuggestionsText: false,
+    suggestions: this.props.data.items || [],
+    value: ''
+  };
+
+  componentWillMount () {
+    this.updateValue(this.props.query[this.props.name] || this.props.value || '');
+  }
+
+  componentWillReceiveProps (nextProps) {
+    if (!this.needsUpdate(nextProps)) return;
+    if (nextProps.error) this.showError();
+    this.updateValue(this.props.query[this.props.name] || nextProps.value || '');
+  }
+
+  onChange = (event, {newValue}) => {
+    this.setState({
+      showNoSuggestionsText: true,
+      value: newValue
+    });
+    this.hideError();
+  }
+
+  onFocus = () => {
+    this.scrollToElement();
+  }
+
+  onBlur = () => {
+    this.hideNoSuggestionsText();
+  }
+
+  onSuggestionsFetchRequested = ({value}) => {
+    this.setState({
+      suggestions: this.getSuggestions(value)
+    });
+  }
+
+  onSuggestionsClearRequested = () => {
+    this.setState({
+      suggestions: []
+    });
+  };
+
+  onSuggestionSelected = (e, {method}) => {
+    this.setState({
+      showNoSuggestionsText: false
+    });
+
+    if (method === 'enter' && !this.props.submitOnEnter) {
+      e.preventDefault();
+    }
+  }
+
+  getSuggestions = value => {
+    const inputValue = value.trim().toLowerCase();
+    const inputLength = inputValue.length;
+
+    if (inputValue) {
+      if (this.props.onGetSuggestions) {
+        this.props.onGetSuggestions(inputValue);
+        return this.props.data.items;
+      }
+
+      return this.props.data.items.filter(data => data.item.toLowerCase().slice(0, inputLength) === inputValue);
+    }
+
+    return [];
+  }
+
+  getSuggestionValue = suggestion => suggestion.item;
+
+  hideNoSuggestionsText = () => {
+    this.setState({
+      showNoSuggestionsText: false
+    });
+  }
+
+  showError () {
+    this.setState({
+      showError: true
+    });
+  }
+
+  hideError () {
+    this.setState({
+      showError: false
+    });
+  }
+
+  // TODO: Move this function into utilities so other components can use it
```

Comment:
Can we do this now? The two possible ways coming to my mind are:

- simply exporting this to a javascript function inside a utils folder;
- creating a higher order component called `Scrollable`.